Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANSILLReduction2.py

```diff
@@ -0,0 +1,952 @@
+            check_distances_match(mtd[solvent_ws], mtd[ws])
+            if self.mode == AcqMode.TOF:
+                # wavelength dependent subtraction, need to rebin
+                solvent_ws_rebin = solvent_ws + '_tr_rebinned'
+                RebinToWorkspace(WorkspaceToRebin=solvent_ws, WorkspaceToMatch=ws,
+                                 OutputWorkspace=solvent_ws_rebin)
+                Minus(LHSWorkspace=ws, RHSWorkspace=solvent_ws_rebin, OutputWorkspace=ws)
+                DeleteWorkspace(solvent_ws_rebin)
+            else:
+                check_wavelengths_match(mtd[solvent_ws], mtd[ws])
+                Minus(LHSWorkspace=ws, RHSWorkspace=solvent_ws, OutputWorkspace=ws)
+
+    def apply_solid_angle(self, ws):
+        '''Calculates solid angle and divides by it'''
+        sa_ws = ws + '_solidangle'
+        # D22B has the front panel tilted, hence the Rectangle approximation is wrong
+        # D16 can be rotated around the sample, where again rectangle is wrong unless we rotate back
+        method = 'GenericShape' if self.instrument == 'D22B' or self.instrument == 'D16' else 'Rectangle'
+        SolidAngle(InputWorkspace=ws, OutputWorkspace=sa_ws, Method=method)
+        Divide(LHSWorkspace=ws, RHSWorkspace=sa_ws, OutputWorkspace=ws, WarnOnZeroDivide=False)
+        DeleteWorkspace(Workspace=sa_ws)
+
+    def apply_masks(self, ws):
+        '''
+        Applies default (edges and permanently bad pixels) and beam stop masks
+        Masks can be created in many ways: manually, via MaskBTP, etc.
+        When creating a mask with instrument viewer, there are 2 possibilities:
+        o) hitting the button Apply and Save as > Detector Mask to Workspace
+            This is the legacy way of propagating the masks.
+            In this case the workspace is a SpecialWorkspace2D which just contains 1s as intensity for the detector that is masked
+        o) hitting the button Apply to Data
+            In this case the workspace has a proper mask.
+        '''
+        edge_mask = self.getProperty('DefaultMaskWorkspace').value
+        if edge_mask and (isinstance(edge_mask, SpecialWorkspace2D) or edge_mask.detectorInfo().hasMaskedDetectors()):
+            MaskDetectors(Workspace=ws, MaskedWorkspace=edge_mask)
+        beam_stop_mask = self.getProperty('MaskWorkspace').value
+        if beam_stop_mask and (isinstance(beam_stop_mask, SpecialWorkspace2D) or beam_stop_mask.detectorInfo().hasMaskedDetectors()):
+            MaskDetectors(Workspace=ws, MaskedWorkspace=beam_stop_mask)
+
+    def apply_thickness(self, ws):
+        '''Normalises by sample thickness'''
+        thicknesses = self.getProperty('SampleThickness').value
+        length = len(thicknesses)
+        if length == 1 and thicknesses[0] > 0:
+            Scale(InputWorkspace=ws, Factor=1/thicknesses[0], OutputWorkspace=ws)
+        else:
+            if thicknesses[0] < 0:
+                thicknesses = mtd[ws].getRun()['SampleThickness'].value.split(',')
```

Comment:
I think the code needs to check that there is an entry called 'SampleThickness' in the sample logs. If I specify SampleThicknessFrom Nexus or enter a negative sample thickness in the UI I can make this line generate an error

Reply:
Added a check for this.